internal/porchutil/function.go

```diff
@@ -0,0 +1,155 @@
+// Experimental: This logic may change to only if all function keywords are found from  expectedKeywords,
+// can it claims a match (return true).
+func ContainsKeywords(function v1alpha1.Function, expectedKeywords []string) bool {
+	if len(expectedKeywords) == 0 {
+		// Accept all functions if keywords are not given.
+		return true
+	}
+	for _, actual := range function.Spec.Keywords {
+		for _, expected := range expectedKeywords {
+			if actual != expected {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// ListKeywords returns all the unique keywords from Porch Functions.
+func (p *PorchFunctionClient) ListKeywords(cmd *cobra.Command) ([]string, error) {
+	keywordsMap := map[string]bool{}
+	fnType := cmd.Flag("type").Value.String()
+	for _, function := range p.Functions {
+		if !IsFunctionTypeMatch(function, fnType) {
+			continue
+		}
+		for _, keyword := range function.Spec.Keywords {
+			keywordsMap[keyword] = true
+		}
+	}
+	var keywords []string
+	for keyword := range keywordsMap {
+		keywords = append(keywords, keyword)
+	}
+	return keywords, nil
+}
+
+func getKeywordsFromFlag(cmd *cobra.Command) []string {
+	flagVal := cmd.Flag("keywords").Value.String()
+	flagVal = strings.TrimPrefix(flagVal, "[")
+	flagVal = strings.TrimSuffix(flagVal, "]")
+	splitted := strings.Split(flagVal, ",")
+	var trimmed []string
+	for _, val := range splitted {
+		trimmed = append(trimmed, strings.TrimSpace(val))
+	}
+	return trimmed
+}
+
+// GetFunctionShortName trim the function image to remove the OCI repo prefix and
+// only show the actual image and tag (or digest).
+// TODO(yuwenma): or may users prefer the Function `meta.name`?
+func GetFunctionShortName(function v1alpha1.Function) string {
+	slash := strings.LastIndex(function.Spec.Image, "/")
+	return function.Spec.Image[slash+1:]
+}
+
+// ListFunctions looks for all the Porch functions which match the command `keywords` and
+// `type` flag.
+func (p *PorchFunctionClient) ListFunctions(cmd *cobra.Command) ([]string, error) {
+	var suggestedFunctions []string
+	fnType := cmd.Flag("type").Value.String()
+	keywords := getKeywordsFromFlag(cmd)
+	for _, function := range p.Functions {
+		if !IsFunctionTypeMatch(function, fnType) {
+			continue
+		}
+		if !ContainsKeywords(function, keywords) {
+			continue
+		}
+		suggestedFunctions = append(suggestedFunctions, GetFunctionShortName(function))
+	}
+	return suggestedFunctions, nil
+}
+
+func visitFunctions() ([]v1alpha1.Function, error) {
+	kubeconfig, ok := os.LookupEnv("KUBECONFIG")
+	if !ok {
+		kubeconfig = filepath.Join(
+			os.Getenv("HOME"), ".kube", "config",
+		)
+	}
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var functions v1alpha1.FunctionList
+	raw, err := clientset.RESTClient().Get().Timeout(PorchExpiration).AbsPath(functionAPI).DoRaw(context.TODO())
```

Comment:
interesting ... this will list all functions in all namespaces in the k8s cluster.
not something to figure out now but the namespace-deness of k8s is making things less straightforward since it is unclear which namespace we'd be querying if we had to pick.

So this is fine for now.

Reply:
agreed. let's keep this; for autocomplete being broader is good, and we can adjust in the future.